datafusion/physical-expr/src/aggregate/median.rs

```diff
@@ -111,13 +112,28 @@ impl PartialEq<dyn Any> for Median {
 /// The intermediate state is represented as a List of those scalars
 struct MedianAccumulator {
     data_type: DataType,
+    batches: Vec<ArrayRef>,
     all_values: Vec<ScalarValue>,
 }
 
+fn to_scalar_values(arrays: &[ArrayRef]) -> Result<Vec<ScalarValue>> {
+    let num_values: usize = arrays.iter().map(|a| a.len()).sum();
+    let mut all_values = Vec::with_capacity(num_values);
+
+    for array in arrays {
+        for index in 0..array.len() {
+            all_values.push(ScalarValue::try_from_array(&array, index)?);
+        }
+    }
+
+    Ok(all_values)
+}
+
 impl Accumulator for MedianAccumulator {
     fn state(&self) -> Result<Vec<ScalarValue>> {
-        let state =
-            ScalarValue::new_list(Some(self.all_values.clone()), self.data_type.clone());
+        let all_values = to_scalar_values(&self.batches)?;
+        let state = ScalarValue::new_list(Some(all_values), self.data_type.clone());
```

Comment:
I wonder if we can change `ScalarValue::List` to use `ArrayRef` instead of `Vec<ScalarValue>` internally.
This would avoid quite some expensive conversions from/to `ScalarValue`.

Reply:
Yes that would help a lot, maybe we can add a `ScalarValue::Array` variant to `ScalarValue`.

Reply:
Maybe @Dandandan  was suggesting

```rust
impl ScalarValue {
  ...
  List(ArrayRef)
  ...
}
```

In general this would work well with the approach @tustvold  is working on upstream in arrow-rs with `Datum` -- https://github.com/apache/arrow-rs/pull/4393

Reply:
Yeah that's what I was trying to suggest - this would avoid the need to convert / allocate to individual `ScalarValue`s and convert to `Array` later on again.

Reply:
That would work well, the reason I was suggesting adding an `Array` variant instead of changing `List` was to avoid changing all the code that depends on `List(Option<Vec<ScalarValue>>, FieldRef)`, but yes long term that would probably be best.

Reply:
After thinking about this some more, I think the most performant thing to do will be to implement a native `GroupsAccumulator` (aka #6800 ) for median. With sufficient effort we could make median be very fast -- so I think this is a good improvement for now